libcontainer/container_linux.go

```diff
@@ -212,21 +212,46 @@ func (c *Container) Start(process *Process) error {
 // after start returns but opens the fifo after start returns.
 func (c *Container) Run(process *Process) error {
 	c.m.Lock()
-	defer c.m.Unlock()
 	if err := c.start(process); err != nil {
+		c.m.Unlock()
 		return err
 	}
 	if process.Init {
-		return c.exec()
+		err := c.exec()
+		if err != nil {
+			c.m.Unlock()
```

Comment:
Why not using `defer c.m.Unlock()` instead of adding unlock operation everywhere?